objstore/s3: make IdleConnTimeout in http.Transport configurable

GiedriusS opened the pull request:
Let's make the http.Transport used inside objstore/s3 configurable. It helps in some cases where a HAProxy in front of the S3 storage sometimes has smaller time-out values and then spurious messages are printed to the logs which are actually harmless. It seems that it also helps with invalid HTTP queries that Go's stdlib sometimes sends to check if a connection is still alive.

## Changes

Added new fields to the Config struct and added the default values from the code before. No impact to the user at all if they have nothing specified in those config files.

The documentation should be updated as well but I am not sure what kind of format should I follow for the docs so I left it up to you :)

## Verification

Added unit tests for checking the new parsing function which has default values.

https://github.com/minio/minio-go/issues/1037 check this link for more information.

GiedriusS commented on pkg/objstore/s3/s3.go:
```diff
@@ -40,13 +40,20 @@ const DirDelim = "/"
+	SignatureV2           bool   `yaml:"signature-version2"`
+	SSEEncryption         bool   `yaml:"encrypt-sse"`
+	SecretKey             string `yaml:"secret-key"`
+	IdleConnTimeout       int    `yaml:"idle-conn-timeout`
```
Don't you mean `model.Duration`? It seems like `model.Time` can only unmarshal strings of format "a.b" and then it converts it into seconds whereas `model.Duration` has a whole parser for different units like you said: https://github.com/prometheus/common/blob/master/model/time.go#L182

I agree with you there, though. Some software only recognizes and accepts seconds however we could be smarter here and re-use some of the Prometheus code to implement it properly :)

bwplotka replied:
I meant Duration, sorry (:

jturolla commented:
What problem are you trying to solve? Do you have issues with the defaults?

We have not encountered issues with this, maybe if you're experiencing timeouts you should move to a bucket near your location?

GiedriusS commented:
Yes, I guess that they probably only make sense when using Google's or Amazon's S3 storage. In our setup, we have a HAProxy in front of a Ceph-based S3 storage. This is related to keepalive connection timeouts mismatches between the HAProxy and the http.Client in s3.go in Thanos. I outlined the issue here: https://github.com/minio/minio-go/issues/1037. I could provide you the exact timeout values but you get the idea: the mismatch causes that harmless message to be printed from time to time, and the implementation of `http.Client` also prints a message that a keep-alive connection had been closed unexpectedly, when that 400 error happens, in stdout.
The bucket is certainly close to the location where Thanos is running, the latencies are around <1ms. :) That's not the problem.

GiedriusS commented:
Sorry for the mess with the commits and rebasing, should be all OK now :)

GiedriusS commented on pkg/objstore/s3/s3.go:
```diff
@@ -60,7 +68,16 @@ type Bucket struct {
 // NewBucket returns a new Bucket using the provided s3 config values.
 func NewBucket(logger log.Logger, conf []byte, reg prometheus.Registerer, component string) (*Bucket, error) {
-	var config Config
+	defIdleConnTimeout, _ := model.ParseDuration("90s")
```
The idea here was to use `model.ParseDuration` so that we would less associate with the actual underlying type of `model.Duration` but I guess I will change it to just use `90 * time.Second` if that is OK with you. I agree that it definitely makes things clearer. As for other points, I agree. I thought it was the default and only acceptable behaviour of unmarshal function but we can move it into a function - it would be much nicer.

bwplotka replied:
:+1:

GiedriusS commented:
In our particular case only `IdleConnTimeout` is enough but what about the other options? Perhaps someone may want to change the other ones as well so I added these to this PR just in case. I don't quite agree with adding one single option for tweaking the idle connections timeout because e.g. other storage providers may not use at all the HTTP protocol underneath their calls.
I think that having one single defaults like this isn't really possible because someone might have e.g. different timeout settings in their haproxy instances and it might be impossible to change on them on the server itself if, for example, it is used intensively by other applications.
I'll make changes and update the PR. Thank you for your review.

bwplotka commented:
> Perhaps someone may want to change the other ones as well so I added these to this PR just in case. 
 
Yes, but it's not a problem to add this later if needed. Those options are quite low level and requires deep knowledge on HTTP client library trivia. I would vote for adding what you need and also pack it in some `http-config` nested structure, so it will look like:

```
type: S3
config:
  http_config:
    idle_conn_timeout: <val>
```
What do you think?

bwplotka commented on pkg/objstore/s3/s3.go:
```diff
@@ -40,13 +41,20 @@ const DirDelim = "/"
+	SignatureV2           bool           `yaml:"signature-version2"`
+	SSEEncryption         bool           `yaml:"encrypt-sse"`
+	SecretKey             string         `yaml:"secret-key"`
+	IdleConnTimeout       model.Duration `yaml:"idle-conn-timeout"`
```
Can we do yaml names with underscore? I don't know how we slipped the `dash` convention, but I would stay with the style Prometheus has: https://prometheus.io/docs/prometheus/latest/configuration/configuration/#%3Cscrape_config%3E

GiedriusS commented:
@bwplotka could you PTAL again? I moved the HTTP config to a different struct and the parsing to a separate function, added a simple unit test for it. I still think there is value in being able to configure other values as well. Being able to configure idle connection timeouts works for us here but someone might run into other issues if they have other settings in their reverse proxy.

GiedriusS commented:
@bwplotka ping.

bwplotka commented on pkg/objstore/s3/s3.go:
```diff
@@ -31,13 +32,25 @@ const DirDelim = "/"
+	HTTPConfig    HTTPConfig `yaml:"http_config"`
+}
+
+// HTTPConfig stores the http.Transport configuration for the s3 minio client
```
missing trailing period

bwplotka commented on pkg/objstore/s3/s3.go:
```diff
@@ -48,10 +61,24 @@ type Bucket struct {
 	sse    encrypt.ServerSide
 }
 
+// ParseConfig unmarshals a buffer into a Config with default HTTPConfig values
```
missing trailing period

bwplotka commented on pkg/objstore/s3/s3.go:
```diff
@@ -48,10 +61,24 @@ type Bucket struct {
 
+// ParseConfig unmarshals a buffer into a Config with default HTTPConfig values
+func ParseConfig(conf []byte) (Config, error) {
+	defaultHTTPConfig := HTTPConfig{IdleConnTimeout: model.Duration(90 * time.Second),
```
Let's format it nicely. We need to put newline for each field as the number of elements is large.

bwplotka commented on pkg/objstore/s3/s3.go:
```diff
@@ -31,13 +32,25 @@ const DirDelim = "/"
+
+// HTTPConfig stores the http.Transport configuration for the s3 minio client
+type HTTPConfig struct {
+	IdleConnTimeout       model.Duration `yaml:"idle_conn_timeout"`
```
I would really suggest just doing the thing you need right now which is IdleConnTimeout explicitly for HA Proxy.

I feel we should stick to YAGNI here and not introduce something that is not useful right now: https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it

GiedriusS commented:
Updated the PR according to your comments. I don't see a problem with giving the users complete power in terms of configuration to the users but I understand your concerns and I only left `idle_conn_timeout` for now since that is what is causing these problems in the first place. Please take a look when you have time.
